qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can
+    handle either a Pauli operator one would like to measure or a measurement on a set
+    of qubits.
+    """
+
+    def __init__(self, observable: Pauli | None = None):
+        """
+        LightCone initializer.
+
+        Args:
+            observable: If None the lightcone will be computed for the set
+            of measurements in the circuit.
+            If a Pauli operator is specified, the lightcone will correspond to
+            the reduced circuit with the
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
+        """For now this method assumes that all the measurements in the circuit are
+        final measurements."""
+        qubits_measured = set()
+        for node in dag.topological_nodes():
+            # print(getattr(node, "name", False))
+            if getattr(node, "name", False) == "measure":
+                qubits_measured |= {x for x in node.qargs}
+        print(qubits_measured)
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[int], tuple[Instruction, list[int]]]:
+        """Returns the initial lightcone.
+        If obervable is None, the lightcone is a different Z operator in each one of the
+        qubits with a measurement. If a Pauli observable is defined, that will be
+        the first gate in the circuit.
+        """
+        if self.observable is None:
+            qubits_measured = self._find_measurement_qubits(dag)
+            light_cone = qubits_measured
+            light_cone_ops = [(PauliGate("Z"), [qubit_index]) for qubit_index in qubits_measured]
+        else:
+            pauli_string = str(self.observable)
+            light_cone = [dag.qubits[i] for i, p in enumerate(pauli_string) if p != "I"]
+            light_cone_ops = [(PauliGate(pauli_string.replace("I", "")), light_cone)]
+
+        return set(light_cone), light_cone_ops
+
+    def run(self, dag: DAGCircuit):
+        """Run the LightCone pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+        light_cone, light_cone_ops = self._get_initial_lightcone(dag)
+
+        print(light_cone, light_cone_ops)
+        # iterate over the nodes in reverse topological order
+        for node in reversed(list(dag.topological_op_nodes())):
+            if not light_cone.intersection(node.qargs):
+                dag.remove_op_node(node)
+                continue
+
+            commutes_bool = True
+            for op in light_cone_ops:
+                commute_bool = commutator.commute(op[0], op[1], [], node.op, node.qargs, [])
```

Comment:
We should handle the `max_num_qubits` arg, which currently defaults to 3. E.g. if a user would give a `PauliGate("ZZZZ")` I think the code would just forego the commutation check and return `False` for any commutation.

So we should probably set `max_num_qubits=max(len(op[1]), len(node.qargs))` and add a warning if that number becomes too big.